### PR TITLE
Support blocking queries

### PR DESCRIPTION
**Pull request by ryanuber**
This adds blocking queries to all of the RPC endpoints. This transitively allows blocking queries to be supported by all of the HTTP endpoints, except the /v1/agent/* ones. All of the blocking queries are scoped to minimize waking clients when there is no data change, so for example a blocking query on /v1/node/node1/allocations will not fire if a new allocation is made on node2.

### CODE REVIEW

File: nomad/node_endpoint.go [hunk 1]
```diff
@@ -330,9 +339,9 @@ func (n *Node) GetAllocs(args *structs.NodeSpecificRequest,
 
 	// Setup the blocking query
 	opts := blockingOptions{
-		queryOpts:  &args.QueryOptions,
-		queryMeta:  &reply.QueryMeta,
-		allocWatch: args.NodeID,
+		queryOpts: &args.QueryOptions,
+		queryMeta: &reply.QueryMeta,
+		watch:     watch.NewItems(watch.Item{AllocNode: args.NodeID}),
```
Comment (ryanuber) [inline nomad/node_endpoint.go]: I wasn't sure if we would want to also watch the node entry here. It's possible to do that with the framework used here, but we would increase the frequency this watch would fire for potentially no change.

**Comment by armon** [comment]
@ryanuber Wow, you went big! I'm wondering, is there any commonality between how we do it here, and Consul? I'm wondering if we can have a shared "watch" lib between the two. Potentially makes sense to add hooks into MemDB to support this as well.

Thoughts? /cc: @slackpad

**Comment by sethvargo** [comment]
:heart:

**Comment by ryanuber** [comment]
There is definitely common functionality here which we can abstract and share between both Consul and Nomad. MemDB would be a good place to expose the low-level change notification, and we can just implement the delta watch triggering in each respective project we need it for! Sounds great to me, this will just be for 0.2 or until we find a good way of exposing the functionality in a way that's useful for the greater good.

**Comment by slackpad** [comment]
Totally agree we could share this and MemDB seems like the right place. If MemDB handles table-level triggering with some data to inspect about what changed then the project-level stuff I think would be filtering and aggregation. Like MemDB calls a registered hook and gives you a list of keys added/removed or something like that. Would need to think on it a little more.

For Consul making the watches more granular would be the perfect time to do this.

I like how @ryanuber you've exposed some pretty low-level stuff up to the RPC call site to let the caller pick what they want. I kind of went back and forth on this and ended up having a mapping to state_store functions (kind of like the pre-MemDB code) but this way is probably better.

**Comment by dadgar** [comment]
LGTM

**Pull request merged**